pandas/core/algorithms.py

```diff
@@ -1829,11 +1830,25 @@ def diff(arr, n: int, axis: int = 0):
     -------
     shifted
     """
+    from pandas.core.arrays import PandasDtype
 
     n = int(n)
     na = np.nan
     dtype = arr.dtype
 
+    if dtype.kind == "b":
+        op = operator.xor
+    else:
+        op = operator.sub
+
+    if isinstance(dtype, PandasDtype):
+        # PandasArray cannot necessarily hold shifted versions of itself.
+        arr = np.asarray(arr)
+        dtype = arr.dtype
```

Comment:
would it be more idiomatic to do extract_array up front?

Reply:
I would say not because we don't need to "extract" the array, since arrays are already passed (this doesn't get passed Series or Index objects). You can of course use `extract_array` to get rid of PandasArrays, but I think the above is more explicit.